src/Authentication/Authentication/TokenCache/MacTokenCache.cs

```diff
@@ -0,0 +1,214 @@
+                    keychainOrArray: IntPtr.Zero,
+                    serviceNameLength: (uint)Constants.TokenCahceServiceName.Length,
+                    serviceName: Constants.TokenCahceServiceName,
+                    accountNameLength: (uint)appId.Length,
+                    accountName: appId,
+                    passwordLength: out uint passwordLength,
+                    passwordData: out passwordDataPtr,
+                    itemRef: out itemPtr);
+
+                if (resultStatus == MacNativeKeyChain.SecResultCodes.errSecItemNotFound)
+                    return contentBuffer;
+
+                else if (resultStatus != MacNativeKeyChain.SecResultCodes.errSecSuccess)
+                {
+                    throw new Exception(string.Format(
+                        CultureInfo.CurrentCulture,
+                        ErrorConstants.Message.MacKeyChainFailed,
+                        "SecKeychainFindGenericPassword",
+                        resultStatus));
+                }
+
+                if (itemPtr != IntPtr.Zero && passwordLength > 0)
+                {
+                    contentBuffer = new byte[passwordLength];
+                    Marshal.Copy(passwordDataPtr, contentBuffer, 0, contentBuffer.Length);
+                }
+                return contentBuffer;
+            }
+            finally
+            {
+                FreePointers(ref itemPtr, ref passwordDataPtr);
+            }
+        }
+
+        /// <summary>
+        /// Adds or updates an app's token to MacOS KeyChain.
+        /// </summary>
+        /// <param name="appId">An app/client id.</param>
+        /// <param name="plainContent">The content to store.</param>
+        public static void SetToken(string appId, byte[] plainContent)
+        {
+            IntPtr passwordDataPtr = IntPtr.Zero;
+            IntPtr itemPtr = IntPtr.Zero;
+            try
+            {
+                int resultStatus = MacNativeKeyChain.SecKeychainFindGenericPassword(
+                    keychainOrArray: IntPtr.Zero,
+                    serviceNameLength: (uint)Constants.TokenCahceServiceName.Length,
+                    serviceName: Constants.TokenCahceServiceName,
+                    accountNameLength: (uint)appId.Length,
+                    accountName: appId,
+                    passwordLength: out uint passwordLength,
+                    passwordData: out passwordDataPtr,
+                    itemRef: out itemPtr);
+
+                if (resultStatus != MacNativeKeyChain.SecResultCodes.errSecSuccess &&
+                    resultStatus != MacNativeKeyChain.SecResultCodes.errSecItemNotFound)
+                {
+                    throw new Exception(string.Format(
+                        CultureInfo.CurrentCulture,
+                        ErrorConstants.Message.MacKeyChainFailed,
+                        "SecKeychainFindGenericPassword",
+                        resultStatus));
+                }
+
+                if (itemPtr != IntPtr.Zero)
```

Comment:
>itemPtr != IntPtr.Zero [](start = 20, length = 22)

Q: is this equivalent to resultStatus == MacNativeKeyChain.SecResultCodes.errSecSuccess? If so, it can be a good idea to change the result evaluation into a switch where the cases are self documenting: i.e. errSecItemNotFound will document the "// Key not found" message below.

Reply:
You are right. `itemPtr` is a pointer to a reference to the new keychain item. Thus, if a keychain item is found, `itemPtr` will point to a reference; if not, it will be null. I'll simplify this evaluation as per your suggestion.